saleor/graphql/account/bulk_mutations/customer_bulk_update.py

```diff
@@ -0,0 +1,598 @@
+                customer_input,
+                index,
+                index_error_map,
+            )
+
+            if new_external_ref in duplicated_refs:
+                message = "Duplicated externalReference value."
+                index_error_map[index].append(
+                    CustomerBulkUpdateError(
+                        path="input.externalReference",
+                        message=message,
+                        code=CustomerBulkUpdateErrorCode.DUPLICATED_INPUT_ITEM.value,
+                    )
+                )
+                base_error_count += 1
+
+            if shipping_address_data:
+                clean_shipping_address = cls.clean_address(
+                    shipping_address_data,
+                    address_type=AddressType.SHIPPING,
+                    field=SHIPPING_ADDRESS_FIELD,
+                    index=index,
+                    index_error_map=index_error_map,
+                )
+                customer_input["input"][SHIPPING_ADDRESS_FIELD] = clean_shipping_address
+
+            if billing_address_data:
+                clean_billing_address = cls.clean_address(
+                    billing_address_data,
+                    address_type=AddressType.BILLING,
+                    field=BILLING_ADDRESS_FIELD,
+                    index=index,
+                    index_error_map=index_error_map,
+                )
+                customer_input["input"][BILLING_ADDRESS_FIELD] = clean_billing_address
+
+            if email := customer_input.get("email"):
+                customer_input["email"] = email.lower()
+
+                if customer_input["email"] in duplicated_emails:
+                    message = "Duplicated email value."
+                    code = CustomerBulkUpdateErrorCode.DUPLICATED_INPUT_ITEM.value
+                    index_error_map[index].append(
+                        CustomerBulkUpdateError(
+                            path="input.email", message=message, code=code
+                        )
+                    )
+                    base_error_count += 1
+
+            if base_error_count > 0:
+                cleaned_inputs_map[index] = None
+            else:
+                cleaned_inputs_map[index] = customer_input
+
+        return cleaned_inputs_map
+
+    @classmethod
+    def get_customers(cls, cleaned_inputs_map: dict) -> list[models.User]:
+        lookup = Q()
+        for customer_input in cleaned_inputs_map.values():
+            if not customer_input:
+                continue
+
+            single_customer_lookup = Q()
+
+            if customer_id := customer_input.get("id"):
+                single_customer_lookup |= Q(pk=customer_id)
+            else:
+                single_customer_lookup |= Q(
+                    external_reference=customer_input.get("external_reference")
+                )
+            lookup |= single_customer_lookup
+
+        customers = models.User.objects.filter(lookup, is_staff=False)
```

Comment:
Staff can also be a customer; there is `User.objects.customers()` queryset for that.